dataflow/src/ops/rewrite.rs

```diff
@@ -0,0 +1,240 @@
+use std::collections::{HashMap, HashSet};
+use prelude::*;
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Rewrite {
+    src: IndexPair,
+    should_rewrite: IndexPair,
+
+    rw_col: usize,
+    value: DataType,
+
+    key: usize,
+}
+
+impl Rewrite {
+    pub fn new(src: NodeIndex, should_rewrite: NodeIndex, rw_col: usize, value: DataType, key: usize) -> Rewrite {
+        Rewrite {
+            src: src.into(),
+            should_rewrite: should_rewrite.into(),
+            rw_col: rw_col,
+            value: value,
+            key: key,
+        }
+    }
+
+    fn rewrite(&self, r: Vec<DataType>) -> Vec<DataType> {
+        let mut r = r.clone();
+        r[self.rw_col] = self.value.clone();
+        r
+    }
+
+    fn generate_row(&self, r: Vec<DataType>, positive: bool) -> Record {
+        if positive {
+            Record::Positive(r.clone())
+        } else {
+            Record::Negative(r.clone())
+        }
+    }
+}
+
+
+impl Ingredient for Rewrite {
+    fn take(&mut self) -> NodeOperator {
+        Clone::clone(self).into()
+    }
+
+    fn ancestors(&self) -> Vec<NodeIndex> {
+        vec![self.src.as_global(), self.should_rewrite.as_global()]
+    }
+
+    fn is_join(&self) -> bool {
+        true
+    }
+
+
+    fn must_replay_among(&self) -> Option<HashSet<NodeIndex>> {
+        Some(Some(self.src.as_global()).into_iter().collect())
+    }
+
+    fn on_connected(&mut self, _: &Graph) {}
+
+    fn on_commit(&mut self, _: NodeIndex, remap: &HashMap<NodeIndex, IndexPair>) {
+        self.src.remap(remap);
+        self.should_rewrite.remap(remap);
+    }
+
+    fn on_input(
+        &mut self,
+        from: LocalNodeIndex,
+        rs: Records,
+        _: &mut Tracer,
+        _: Option<usize>,
+        nodes: &DomainNodes,
+        state: &StateMap,
+    ) -> ProcessingResult {
+        assert!(from == *self.src || from == *self.should_rewrite);
+        let misses = Vec::new();
+        let mut emit_rs = Vec::with_capacity(rs.len());
+
+        if rs.is_empty() {
+            return ProcessingResult {
+                results: rs,
+                misses: vec![],
+            };
+        }
+
+        for r in rs {
+            if from == *self.src {
+                let key = r[self.key].clone();
+                // ask should_rewrite if column should be rewritten
+                let rc = self.lookup(
+                    *self.should_rewrite,
+                    &[0],
+                    &KeyType::Single(&key),
+                    nodes,
+                    state,
+                ).unwrap();
+
+
+                if rc.is_none() {
+                    // todo (larat): might need to do some stuff here
```

Comment:
This is if you miss in the `should_rewrite` state. You should add that key to `misses`. If you don't that key will never be replayed, and those records will just never be emitted.